### PR TITLE
[workflow] Convert DAG to workflow

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

This PR implements the workflow API that converts Ray DAG into Ray workflow.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
Could you check whether we can use this function instead of writing a recursive by ourself https://github.com/ray-project/ray/blob/c844c706bffd7868fb904384902f3b0b7d25f2a1/python/ray/experimental/dag/dag_node.py#L171

**Comment by iycheng** [comment]
Could you check whether we can use this function instead of writing a recursive by ourself https://github.com/ray-project/ray/blob/c844c706bffd7868fb904384902f3b0b7d25f2a1/python/ray/experimental/dag/dag_node.py#L171

File: python/ray/workflow/dag_to_workflow.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+from ray import workflow
+
+from ray.experimental.dag import DAGNode, FunctionNode
+from ray.experimental.dag.input_node import InputAtrributeNode, DAGInputData
+
+
+def _process_input_value(v: Any, input_context):
+    if isinstance(v, FunctionNode):
+        return _transform_dag_recursive(v, input_context)
```
Comment (iycheng) [inline python/ray/workflow/dag_to_workflow.py:11]: Did we do deduplicate here? For example, if the same inputs are given, will we end up with two workflow steps?

Comment (suquark) [reply python/ray/workflow/dag_to_workflow.py]: good catch!

File: python/ray/workflow/tests/test_dag_to_workflow.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+def test_dag_to_workflow_options(workflow_start_regular_shared):
+    """This test if the workflow inherits DAG options."""
+
+    @ray.remote
+    def no_resource():
+        pass
+
+    # TODO(suquark): The current Ray DAG is buggy, it failed to return the
+    # "original" options, we need to override "num_returns" to pass workflow check.
```
Comment (iycheng) [inline python/ray/workflow/tests/test_dag_to_workflow.py:49]: Do you want to file an issue and fixed it?  @jiaodong did you meet this kind of error before?

**Comment by iycheng** [comment]
Nice work! Do you want to add more tests like make sure deduplicate is working. And also use ObjectRef as the input and make sure it's working?

I think in the end when we deprecate the `step` we probably need to convert some of our existing test cases to the DAG.

File: python/ray/workflow/tests/test_dag_to_workflow.py (hunk 2 above)
Comment (jiaodong) [reply python/ray/workflow/tests/test_dag_to_workflow.py:49]: nope but I also haven't tested num_returns flag before. feel free to file an issue about this

**Comment by suquark** [comment]
I already add assert for the duplication in the code. Testing ObjectRef is a good idea.

**Comment by iycheng** [comment]
@suquark I'm only talking about, input as an object ref. For example, migrate this test (https://github.com/ray-project/ray/blob/master/python/ray/workflow/tests/test_serialization.py#L88)

**Comment by suquark** [comment]
I add more tests for objectrefs

**Review by fishbone — APPROVED** [review]
LGTM

**Pull request merged**